Moved gcc optimization flags into the optimization section.

beyondCritics opened the pull request:
Proposal to fix issue https://github.com/official-stockfish/Stockfish/issues/1164#issuecomment-313495059
f-profile-generate and -fprofile-use builds get the same build flags.
The logic of the make file, as i understand it, dictates that all optimization issues are handled in dedicated section 3.3. Within each section, there is further dispatching according to lower level issues.

MichaelB7 commented:
Although I believed this change was logical , it created a 0.6% slowdown in nps on macOS, PG0 build with gcc4.8, on long concurrent bench with turbo boost off.  

```bench 64 1 24 
===========================
master
Total time (ms) : 294257
Nodes searched  : 521013561
Nodes/second    : 1770607

===========================
test
Total time (ms) : 296113
Nodes searched  : 521013561
Nodes/second    : 1759509
```

MichaelB7 commented:
find script sample here: https://github.com/official-stockfish/Stockfish/pull/1157 scroll down a  a little

just as a follow-up, if I include both flags. i.e., accept your addition  at line 280 and reject your deletion at lines 539/540, that produces the fastest exe for me at about 0.6% inc in speed  - maybe that's what you meant?
'''
 CXXFLAGS += -fno-peel-loops -fno-tracer
'''

mcostalba commented:
Adding very specific optimization compiler flags is very fragile and not future proof because it can (and it does change) from gcc's version to version.

We don't have to deal with inner of gcc, just set usual -O3 or similar general scope optimization flags and that's it: people doing specialized compiling will play the flag game themselves (they enjoy doing it btw), but it is poor choice to tweak the flags ourselves: if we open that door it will be a flag change patch every week because it is too platform / compiler dependent.

So just remove special flags all over and that's it: consistent, easy, future proof, simple.

mcostalba commented:
BTW the flag being added is -fno-peel-loops, it says to gcc to never unroll the loops: this is a silly thing to order to the compiler because how can you know that never unrolling the loop is the best way now and in the future?

You should really **NOT** deal with such low level details and let the compiler to do its job.

beyondCritics commented:
I have changed my proposal according to your wishes.
Since there is a conflict (speed <-> simplicity) i have added a comment, explaining the rationale, in order to save time for other contributors.

mstembera commented:
Removing -fno-peel-loops and -fno-tracer results in an over 3%! slowdown for profile builds as independently confirmed many times in the past.  This has been the case since gcc 4.9.x.  See links at top of https://github.com/official-stockfish/Stockfish/pull/237 for some history.  I have also been confirming this remains the case with each major version of gcc since then(till current 7.1.0) so it's not version specific and highly unlikely to change.  Also note that -fno-peel-loops is NOT the same as -fno-unroll-loops so loops still do get unrolled.
https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html
Since we use PGO on Fishtest removing this also effectively reduces fishtest capacity by 3%.

```
Results for 20 tests for each version:

            Base      Test      Diff      
    Mean    1115266   1079672   35594     
    StDev   13630     14591     2433      

p-value: 0
speedup: -0.032
```

MichaelB7 commented:
@mcostalba  makes the best argument  hands down, following the KISS principle.   Those of us that need a faster compile can always add the flags that work for our individual set up.

ppigazzini commented:
@zamar @mcostalba @mstembera 
big slowdown also on my i7 3770K + WSL: I'm worried for the waste of time&energy in fishtest

gcc version 5.4.1 20160904 (Ubuntu 5.4.1-2ubuntu1~16.04)
6.7% slowdown
```
bash ./bench-parallel.sh ./stockfish-previous ./stockfish-master 30
base =    2448503 +/- 7319
test =    2284202 +/- 6574
diff =    -164301 +/- 1791
speedup = -0.067103
```

gcc version 7.1.0 (Ubuntu 7.1.0-5ubuntu2~16.04)
4.3% slowdown 
```
bash ./bench-parallel.sh ./stockfish-previous ./stockfish-master 30
base =    2472248 +/- 10604
test =    2366550 +/- 9370
diff =    -105698 +/- 1842
speedup = -0.042754
```

ddugovic commented:
Is fishtest consuming more time now?  I thought fishtest games were timed.

I was about to ask whether there was an option to suggest to the compiler to maximize resource usage during building; I now see that `-O3` which includes `-fexpensive-optimizations` already serves that purpose (and is generic/future-proof).
